wrappers/_shared.nix

```diff
@@ -42,18 +43,34 @@ in
 
   config = mkMerge [
     # Make our lib available to the host modules
-    { nixvim.helpers = lib.mkDefault (import ../lib/helpers.nix { inherit pkgs lib; }); }
+    { nixvim.helpers = lib.mkDefault helpers; }
     # Propagate extraFiles to the host modules
     (optionalAttrs (filesOpt != null) (
       mkIf (!cfg.wrapRc) (
         setAttrByPath filesOpt (
           listToAttrs (
             map (
               { target, source, ... }:
+              let
+                needByteCompiling = cfg.performance.byteCompileLua.enable && cfg.performance.byteCompileLua.configs;
```

Comment:
Is this the right place to be doing this? The `_shared.nix` module is only used by the home-manager/NixOS/nix-darwin modules; anything here will not affect the standalone build.

If all builds should be affected, we probably want something in `/modules/top-level`.

(Apologies if I missed the point of the changes, I'm on my phone and haven't looked properly)

Reply:
It is now done with the minimum changes to the options itself (in fact probably no changes at all). So I did it in the final stages just before final files a generated. I found it in two places: first is nixvim itself manages the files in the `nvim-config` plugin. Second place is nixvim is propagating user files to the upper module. For now it's used only with home-manager, because all other modules have `wrapRc = lib.mkForce true;`. User can inspect its config with unmodified files (content, text, source attributes).

Second approach I considered is populating all relevant places in the options itself, but it'll take a refactor of the options. In fact you already done at least part of it in #1889. So when I wrote this I thought that my approach is simpler and less intrusive to nixvim framework. But after your PR it can be re-evaluated, maybe it'll be simpler by doing it _right_.

The following is my thoughts about doing it in options on top of the current main (not your PR), I didn't done it in practice, so maybe I'm overlooking something.

1) Since my `performance` module is in main modules it is populated to submodules (`files`). So we need to make it's default value of `performance.byteCompileLua` to follow top-level value. Can an option in submodule be set with `mkDefault` to the value of top-level module? I'm sure it shouldn't be a problem. User can even disable a particular file from compiling.

2) In the `files` submodule the `plugin` option can be optionally compiled (where `writeLlua`, `writeText` and `writeContent` are).

3) In the top level I didn't find an option with `init.lua` file, it's generated in `let in` block (but your PR fixes that). Also `nixvim-print-init` should be handled (it should print a text file, not the one possibly compiled).

4) All above was easy, but what requires more changes (in the same places as in this PR: `nvim-config` and `shared` module) is `extraFiles`. `files` submodule is also goes there in the end. It has `source` option, but I can't change it, because it can be set by the user. So I need some other option introduced, `finalSource` for example. And it would be populated with source unchanged or compiled depending on the option value and whether the file is lua. `files` submodule also should populate `finalSource`.

5) Maybe `files` submodule shouldn't decide for itself if the file should be compiled. In that case it should populate `source` option in `extraFiles` (as it is no), and `finalPackage` in `extraFiles` is populated depending on the global performance option. But this should be done with care, because top level init.lua module and `files` submodules share the same configs mostly so the differences could be tricky. But maybe neither `init.lua`, nor `files` submodule should do this and it can be left only in `extraFiles`. `init.lua` can be handled specially as it is in this PR.

Reply:
> 1. Can an option in submodule be set with `mkDefault` to the value of top-level module? I'm sure it shouldn't be a problem. User can even disable a particular file from compiling.

This is possible when declaring the submodule type, since submodules can contain `options` `imports` and `config` attrs just like any other module.

You could do:

```nix
lib.types.submodule {
  options = {
    # TODO
  };

  config = {
    # This will inherit the priority from the top-level definition:
    performance.byteCompile = lib.mkDerivedConfig options.performance.byteCompile (v: v);

   # Or this will do the same for all the options in `performance`
   performance = lib.mkAliasAndWrapDefsWithPriority (v: v) options.performance;
  };
}
```

> 2. In the `files` submodule the `plugin` option can be optionally compiled (where `writeLua`, `writeText` and `writeContent` are).

I'm missing the significance of this.

> 3. In the top level I didn't find an option with `init.lua` file, it's generated in `let in` block (but your PR changes that).

If it helps we can look into getting specific parts of that refactor merged separately. That'll also make the refactor easier to test/review since it'll reduce its scope.

> Also `nixvim-print-init` should be handled (it should print a text file, not the one possibly compiled).

What do you mean? Why would this need special handling? Are you pre-compiling the init.lua too?



> * All above was easy, but what requires more changes (in the same places as in this PR: `nvim-config` and `shared` module) is `extraFiles`. `files` submodule is also goes there in the end. It has `source` option, but I can't change it, because it can be set by the user. So I need some other option introduced, `finalSource` for example. And it would be populated with source unchanged or compiled depending on the option value and whether the file is lua. `files` submodule also should populate `finalSource`.
> 
> * Maybe `files` submodule shouldn't decide for itself if the file should be compiled. In that case it should populate `source` option in `extraFiles` (as it does now), and `finalSource` in `extraFiles` is populated depending on the global performance option. But this should be done with care, because top level init.lua module and `files` submodules share the same configs mostly so the differences could be tricky. But maybe neither `init.lua`, nor `files` submodule should do this and it can be left only in `extraFiles`. `init.lua` can be handled specially it `top-level/output.nix` as it is in this PR.

Honestly, I struggled to absorb all of that. But I agree we shouldn't be overriding user-definable options like `extraFiles.*.source` with a compiled version of itself.
I think that'd require infinite recursion anyway.

A `compiledFiles`, `compiledPlugin`, or some other name could certainly be introduced.

---

We also need to watch out for "Import From Derivation" when consuming other derivations (`pkgs.writeText`, produce a derivation).

It's fine to pass a derivation's path into another derivation writer/builder, but we can't import/read the derivation path's _files_ (content) into the nix evaluation itself.

Reply:
> > 3\. In the `files` submodule the `plugin` option can be optionally compiled (where `writeLua`, `writeText` and `writeContent` are).
> 
> I'm missing the significance of this.

Your original point was that I shouldn't done anything in _shared.nix. The only way I know to _not doing this in shared.nix_, is to have all the files generated in files and/or extraFiles submodules. All files should already be compiled properly and separately (without any filePlugin). And if I understand this right, the way files are propagated to extraFiles right now is that `source` of extraFiles is set to `plugin` files. So my idea was, that I just change plugin to write a compiled file and it would automatically get to extraFiles, and that to home-manager.

> > Also `nixvim-print-init` should be handled (it should print a text file, not the one possibly compiled).
> 
> What do you mean? Why would this need special handling? Are you pre-compiling the init.lua too?

If a top-level module has some e.g. `finalSource` option which would have a path to a compiled init.lua, it means that `nixvim-print-init` need to be pointed to another uncompiled copy of the init.lua. That's all I meant.

> We also need to watch out for "Import From Derivation" when consuming other derivations (`pkgs.writeText`, produce a derivation).

Ah I did it because my library function writeByteCompiledLua takes string and I only have generated file in that place. I though that it's alright.

Main question: are you planing to finish your refactor PR anytime soon? If I understand correctly, my current implementation is unsuitable (although it works fine for me). If I need to rewrite it, I need to know what is the state of that refactor PR, because we touch the same places.

Reply:
> Main question: are you planing to finish your refactor PR anytime soon?

I'd like to, but we have higher priorities currently (e.g. the CI being broken, which blocks many PRs including this).

I also don't know what the other maintainers' views are on that refactor PR, like any PR it'll need their approval before merging.

As I said, if there's specific features from that PR (or features related to it) that would help you out implementing this, I could look into making a smaller PR that adds just those.

As I understand it, this PR is also blocked by #1886, so I wouldn't rush to rewrite stuff until the other blockers are resolved.